### ISSUE TITLE
Web-based Launcher Alternative

### ISSUE THREAD

**Issue by mottosso** [getavalon/core#342]
### Goal

Eliminate the dependency on Qt for GUIs not running from within a DCC like Maya

### Motivation

Shipping Python and PyQt5 is a pain

### Implementation

I stumbled upon an interesting Python framework that generates an HTML/CSS/JS application from your Python code, working a lot like Qt as a framework, and put this together.

- https://github.com/mottosso/avalon-launcher2

So far, it feels a bit raw, a little early. But the concept is quite neat.

Novel things:

1. Vendorable, cross-platform GUI library without dependencies, nothing compiled
2. Qt-like API
3. Can serve from a central computer, for e.g. a project manager
4. Accessible via your browser

![](https://user-images.githubusercontent.com/2152766/46261996-139db600-c4f3-11e8-9284-c333cd82d848.gif)

**Comment by tokejepsen** [comment]
Did you actually get it to launch a program?

I've previous tried to launch programs from the browser from there are security measures in place that prevents this.

**Comment by mottosso** [comment]
Yes, the example program already works.

The browser calls the Python process which is also acting as the web server, Python is what launcher processes, just like our current launcher.

https://github.com/mottosso/avalon-launcher2/blob/6d9f89f871e5930b88be533c275c69be2d9f3ac7/launcher/__main__.py#L93

**Comment by aardschok** [comment]
How far did you get with the database integration? I would to see how fast this will be on a live one

**Comment by mottosso** [comment]
The database is mocked in that example, you can find it [here](https://github.com/mottosso/avalon-launcher2/blob/master/launcher/__main__.py#L7). It's just a dictionary.

The overhead from a database would be fetching that dictionary.

Have thought some more about this, and the concept is quite sound; the main crux so far is having to open a browser to show the app. Sometimes that's what you want (for e.g. the asset manager, like ftrack) but other times not (i.e. having a little taskbar icon with notifications and status, or just a smaller and separate window). But then I thought, why not make a Chrome/Firefox "app"? That'd enable not only a native-looking application but also a standalone window and taskbar icons etc. etc. It would effectively be identical to a Qt app, with the added benefit of (1) being usable from a browser, e.g. remotely like ftrack and (2) require no additional dependencies for the end-user (they'll already have a browser).

The framework itself is a little early, a little short on features, but the hard part of getting Python and the browser talking is there, the rest is HTML and things I think we could implement ourselves if needed.

**Comment by davidlatwe** [comment]
With this, I think we could even make some remote actions :P

**Comment by davidlatwe** [comment]
> Can serve from a central computer

If it's served from central computer, would this be able to open local DCC tools :O ?

**Comment by mottosso** [comment]
No, DCCs will open where the Python instance is runnnig, so for the launcher it would only make sense to serve it locally. But for the project manager, we could serve remotely, and then tying them together would be a possibility.

E.g. Project Manager is served from a computer A, and accessed from B via the browser. And then B could have a Python process of its own that A sends commands to. Because it's all in Python (as opposed to something like node-js) it means we can do stuff on A like..

```python
def on_maya_clicked(self, client_ip):
  self.send_command(client_ip, "subprocess.Popen('maya')")
```

So now we've got a one-in-all project management application, written in Python, accessible from all machines, that also launches Maya!

**Comment by davidlatwe** [comment]
Sounds exciting !

**Comment by davidlatwe** [comment]
So this would most possibly be a new Project manager, and the current Launcher would also be the minion to receive the remote command. Right ?

**Comment by mottosso** [comment]
If it works, then we could strip out a vast majority of the current Launcher code, which mostly involves QML, and use it as a background process, running on each users machine instead. Something that runs on startup, for example, that doesn't need to be visible.

All interaction would happen via the served web application.
